### PR TITLE
make jobdispatcher awaitable

### PR DESCRIPTION
**Pull request by TingluoHuang**

### CODE REVIEW

File: src/Agent.Listener/Agent.cs [hunk 1]
```diff
@@ -164,48 +164,58 @@ void CtrlCHandler(object sender, EventArgs e)
...
-                                message = null;
-                            }
+                            var newJobMessage = JsonUtility.FromString<JobRequestMessage>(message.Body);
+                            jobDispatcher.Run(newJobMessage);
+                        }
+                        else if (string.Equals(message.MessageType, JobCancelMessage.MessageType, StringComparison.OrdinalIgnoreCase))
+                        {
+                            var cancelJobMessage = JsonUtility.FromString<JobCancelMessage>(message.Body);
+                            bool findJobToCancel = jobDispatcher.Cancel(cancelJobMessage);
+                            skipMessageDeletion = autoUpdateInProgress && !findJobToCancel;
```
Comment (patrickcarnahan) [inline src/Agent.Listener/Agent.cs]: Maybe rename the variable to jobCancelled or jobFound. findJobToCancel sounds like the caller now needs to go find the job. :-)

File: src/Agent.Listener/JobDispatcher.cs [hunk 2]
```diff
@@ -1,44 +1,231 @@
...
-        Task<int> RunAsync(JobRequestMessage message, CancellationToken jobRequestCancellationToken);
+        void Run(JobRequestMessage message);
+        bool Cancel(JobCancelMessage message);
+        Task WaitAsync(CancellationToken token);
+        Task Shutdown();
     }
 
     public sealed class JobDispatcher : AgentService, IJobDispatcher
     {
+        private class Dispatch : IDisposable
```
Comment (patrickcarnahan) [inline src/Agent.Listener/JobDispatcher.cs]: Maybe call this Job or JobInfo? Dispatch is a verb.

File: src/Agent.Listener/JobDispatcher.cs [hunk 3]
```diff
@@ -1,44 +1,231 @@
...
+        {
+            Trace.Info("Job request {0} received.", jobRequestMessage.JobId);
+
+            Dispatch currentDispatch = null;
+            if (_workerDispatchQueue.Count > 0)
+            {
+                currentDispatch = _workerDispatchQueue.Dequeue();
+            }
+
+            Dispatch newDisptch = new Dispatch(jobRequestMessage.JobId, jobRequestMessage.RequestId);
```
Comment (patrickcarnahan) [inline src/Agent.Listener/JobDispatcher.cs]: Variable is missing an 'a'. newDispatch.

File: src/Agent.Listener/JobDispatcher.cs [hunk 4]
```diff
@@ -1,44 +1,231 @@
...
+
+            newDisptch.WorkerDispatch = RunAsync(jobRequestMessage, currentDispatch, newDisptch.WorkerCancellationTokenSource.Token);
+            _workerCancellationTokenSources.TryAdd(newDisptch.JobId, newDisptch.WorkerCancellationTokenSource);
```
Comment (patrickcarnahan) [inline src/Agent.Listener/JobDispatcher.cs]: Why not use a ConcurrentDictionary<Guid, Dispatch> instead? There is no reason to separate the cancellation token source from the thing which actually owns it.

File: src/Agent.Listener/JobDispatcher.cs [hunk 5]
```diff
@@ -1,44 +1,231 @@
...
+            public Guid JobId { get; }
+            public Task WorkerDispatch { get; set; }
+            public CancellationTokenSource WorkerCancellationTokenSource { get; private set; }
+        }
+
         private static readonly string _workerProcessName = $"Agent.Worker{IOUtil.ExeExtension}";
 
+        private int _poolId;
+
+        private readonly Queue<Dispatch> _workerDispatchQueue = new Queue<Dispatch>();
```
Comment (patrickcarnahan) [inline src/Agent.Listener/JobDispatcher.cs]: You usages of queue seem to have the potential to be used in a non-thread-safe manner.

File: src/Agent.Listener/Agent.cs [hunk 6]
```diff
@@ -161,49 +161,57 @@ private void CtrlCHandler(object sender, EventArgs e)
...
-                            }
+                            var cancelJobMessage = JsonUtility.FromString<JobCancelMessage>(message.Body);
+                            bool jobCancelled = jobDispatcher.Cancel(cancelJobMessage);
+                            skipMessageDeletion = autoUpdateInProgress && !jobCancelled;
                         }
-                        finally
+                    }
+                    finally
+                    {
+                        if (!skipMessageDeletion)
```
Comment (ericsciple) [inline src/Agent.Listener/Agent.cs]: still have null ref chance.

File: src/Agent.Listener/Agent.cs [hunk 7]
```diff
@@ -161,49 +161,57 @@ private void CtrlCHandler(object sender, EventArgs e)
...
                         {
-                            if (message != null)
-                            {
-                                //TODO: make sure we don't mask more important exception
-                                await DeleteMessageAsync(message);
-                                message = null;
-                            }
+                            //TODO: make sure we don't mask more important exception
+                            await DeleteMessageAsync(message);
```
Comment (ericsciple) [inline src/Agent.Listener/Agent.cs]: try catch trace exception

File: src/Agent.Listener/JobDispatcher.cs [hunk 8]
```diff
@@ -1,44 +1,300 @@
...
 {
     [ServiceLocator(Default = typeof(JobDispatcher))]
-    public interface IJobDispatcher : IDisposable, IAgentService
+    public interface IJobDispatcher : IAgentService
     {
-        Task<int> RunAsync(JobRequestMessage message, CancellationToken jobRequestCancellationToken);
+        void Run(JobRequestMessage message);
+        bool Cancel(JobCancelMessage message);
+        Task WaitAsync(CancellationToken token);
+        Task Shutdown();
```
Comment (ericsciple) [inline src/Agent.Listener/JobDispatcher.cs]: ShutdownAsync

File: src/Agent.Listener/JobDispatcher.cs [hunk 9]
```diff
@@ -1,44 +1,300 @@
...
     }
 
+    // This implementation of IDobDispatcher is not thread safe.
+    // It is base on the fact that the current design of agent is dequeue
+    // and process one message from message queue everytime.
+    // In addition, it only execute one job every time, 
+    // and server will not send another job while this one is still running.
     public sealed class JobDispatcher : AgentService, IJobDispatcher
     {
+        private class WorkerDispatcher : AgentService, IDisposable
```
Comment (ericsciple) [inline src/Agent.Listener/JobDispatcher.cs]: move to the bottom.

File: src/Agent.Listener/JobDispatcher.cs [hunk 10]
```diff
@@ -1,44 +1,300 @@
...
+                    // cancel the zombie worker, run next job request.
+                    Trace.Error($"Received job request while previous job {jobDispatch.JobId} still running on worker. Cancel the previous job since the job request have been finished on server side with result: {request.Result.Value}.");
+                    jobDispatch.WorkerCancellationTokenSource.Cancel();
+
+                    // wait 45 sec for worker to finish.
+                    Task completedTask = await Task.WhenAny(jobDispatch.WorkerDispatch, Task.Delay(TimeSpan.FromSeconds(45)));
+                    if (completedTask != jobDispatch.WorkerDispatch)
+                    {
+                        // at this point, the job exectuion might encounter some dead lock and even not able to be canclled.
+                        throw new InvalidOperationException("Job dispatch process has encountered unexpected error, the dispatch task is not able to be canceled within 45 seconds.");
```
Comment (ericsciple) [inline src/Agent.Listener/JobDispatcher.cs]: not localize this exception, should never hitted.

**Comment by ericsciple** [comment]
GREAT!

**Pull request merged**